presto-phoenix/src/main/java/io/prestosql/plugin/phoenix/PhoenixClient.java

```diff
@@ -0,0 +1,453 @@
+        }
+        return format(
+                "UPSERT INTO %s VALUES (%s)",
+                quoted(null, handle.getSchemaName(), handle.getTableName()),
+                params);
+    }
+
+    @Override
+    protected ResultSet getTables(Connection connection, Optional<String> schemaName, Optional<String> tableName)
+            throws SQLException
+    {
+        return super.getTables(connection, toPhoenixSchemaName(schemaName), tableName);
+    }
+
+    @Override
+    protected String getTableSchemaName(ResultSet resultSet)
+            throws SQLException
+    {
+        String schemaName = resultSet.getString("TABLE_SCHEM");
+        return schemaName == null ? DEFAULT_SCHEMA : schemaName;
+    }
+
+    protected static ResultSet getColumns(JdbcTableHandle tableHandle, DatabaseMetaData metadata)
+            throws SQLException
+    {
+        Optional<String> escape = Optional.ofNullable(metadata.getSearchStringEscape());
+        return metadata.getColumns(
+                tableHandle.getCatalogName(),
+                // passing null for schema returns tables in all schemas, so we need to specify the default empty schema name
+                escapeNamePattern(Optional.ofNullable(tableHandle.getSchemaName()), escape).orElse(NULL_SCHEMA_NAME),
+                escapeNamePattern(Optional.ofNullable(tableHandle.getTableName()), escape).orElse(null),
+                null);
+    }
+
+    @Override
+    public Optional<ColumnMapping> toPrestoType(ConnectorSession session, JdbcTypeHandle typeHandle)
+    {
+        switch (typeHandle.getJdbcType()) {
+            case VARCHAR:
+            case NVARCHAR:
+            case LONGVARCHAR:
+            case LONGNVARCHAR:
+                if (typeHandle.getColumnSize() == 0) {
+                    return Optional.of(varcharColumnMapping(createUnboundedVarcharType()));
+                }
+                return super.toPrestoType(session, typeHandle);
+            // TODO add support for TIMESTAMP after Phoenix adds support for LocalDateTime
+            case TIMESTAMP:
+            case TIME_WITH_TIMEZONE:
+            case TIMESTAMP_WITH_TIMEZONE:
+                return Optional.empty();
+            case FLOAT:
+                return Optional.of(realColumnMapping());
+            case ARRAY:
+                JdbcTypeHandle elementTypeHandle = getArrayElementTypeHandle(session, typeHandle);
+                if (elementTypeHandle.getJdbcType() == Types.VARBINARY) {
+                    return Optional.empty();
+                }
+                return toPrestoType(session, elementTypeHandle)
+                        .map(elementMapping -> {
+                            ArrayType prestoArrayType = new ArrayType(elementMapping.getType());
+                            String jdbcTypeName = elementTypeHandle.getJdbcTypeName()
+                                    .orElseThrow(() -> new PrestoException(PHOENIX_METADATA_ERROR, "Type name is missing: " + typeHandle));
```

Comment:
The `toString()` for `JdbcTypeHandle` is not friendly for end users. We might want to improve this if we expect that this can happen in practice.